File: sdk/go/agent.go
// Package atradio is the official Go SDK for atradio.fm, built on the Bluesky
// indigo SDK (github.com/bluesky-social/indigo). It mirrors the Rust and
// TypeScript SDKs: an Agent with high-level record verbs plus a read-only
// AppView client.
package atradio

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/bluesky-social/indigo/xrpc"
)

// LoginOptions configures an app-password login.
type LoginOptions struct {
	// Service is the account's PDS (or an entryway like https://bsky.social).
	Service string
	// Identifier is a handle, DID, or email.
	Identifier string
	Password   string
	// AppView base URL (empty → DefaultAppView).
	AppView string
}

// Agent is the atradio.fm client: an authenticated indigo xrpc.Client plus a
// read-only AppView. Its methods serialize behind a mutex so refresh-token
// rotation can't race (a double-spent refresh token gets the session revoked).
type Agent struct {
	xrpc    *xrpc.Client
	did     string
	AppView *AppView

	mu sync.Mutex
}

type sessionResponse struct {
	AccessJwt  string `json:"accessJwt"`
	RefreshJwt string `json:"refreshJwt"`
	Handle     string `json:"handle"`
	Did        string `json:"did"`
}

// Login authenticates with an app password and returns a ready Agent.
func Login(ctx context.Context, opts LoginOptions) (*Agent, error) {
	host := opts.Service
	if host == "" {
		host = "https://bsky.social"
	}
	client := &xrpc.Client{Host: strings.TrimRight(host, "/")}
	var out sessionResponse
	err := client.Do(ctx, xrpc.Procedure, "application/json", "com.atproto.server.createSession", nil,
		map[string]any{"identifier": opts.Identifier, "password": opts.Password}, &out)
	if err != nil {
		return nil, fmt.Errorf("createSession: %w", err)
	}
	client.Auth = &xrpc.AuthInfo{
		AccessJwt:  out.AccessJwt,
		RefreshJwt: out.RefreshJwt,
		Handle:     out.Handle,
		Did:        out.Did,
	}
	return &Agent{xrpc: client, did: out.Did, AppView: NewAppView(opts.AppView)}, nil
}

// FromClient wraps an already-authenticated indigo xrpc.Client (its Auth.Did is
// used as the acting repo).
func FromClient(client *xrpc.Client, appview string) *Agent {
	did := ""
	if client.Auth != nil {
		did = client.Auth.Did
	}
	return &Agent{xrpc: client, did: did, AppView: NewAppView(appview)}
}

// Did returns the acting account's DID.
func (a *Agent) Did() string { return a.did }

// RefreshSession refreshes the access token using the refresh token and
// persists the rotated tokens on the client. Call it on a timer (and once at
// startup) so writes keep working after a long idle. Returns an error only if
// the refresh token itself is dead (re-login needed).
func (a *Agent) RefreshSession(ctx context.Context) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.xrpc.Auth == nil {
		return fmt.Errorf("not authenticated")
	}
	// refreshSession authenticates with the refresh token, not the access token.
	refresher := &xrpc.Client{
		Host: a.xrpc.Host,
		Auth: &xrpc.AuthInfo{AccessJwt: a.xrpc.Auth.RefreshJwt},
	}
	var out sessionResponse
	if err := refresher.Do(ctx, xrpc.Procedure, "", "com.atproto.server.refreshSession", nil, nil, &out); err != nil {
		return fmt.Errorf("refreshSession: %w", err)
	}
	a.xrpc.Auth.AccessJwt = out.AccessJwt
	a.xrpc.Auth.RefreshJwt = out.RefreshJwt
	return nil
}

func nowRFC3339() string { return time.Now().UTC().Format(time.RFC3339) }

// ---- low-level record ops ----

func (a *Agent) putRecord(ctx context.Context, collection, rkey string, record any) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	in := map[string]any{"repo": a.did, "collection": collection, "rkey": rkey, "record": record}
	return a.xrpc.Do(ctx, xrpc.Procedure, "application/json", "com.atproto.repo.putRecord", nil, in, nil)
}

func (a *Agent) createRecord(ctx context.Context, collection string, record any) (string, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	in := map[string]any{"repo": a.did, "collection": collection, "record": record}
	var out struct {
		URI string `json:"uri"`
	}
	if err := a.xrpc.Do(ctx, xrpc.Procedure, "application/json", "com.atproto.repo.createRecord", nil, in, &out); err != nil {
		return "", err
	}
	return out.URI, nil
}

func (a *Agent) deleteRecord(ctx context.Context, collection, rkey string) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	in := map[string]any{"repo": a.did, "collection": collection, "rkey": rkey}
	return a.xrpc.Do(ctx, xrpc.Procedure, "application/json", "com.atproto.repo.deleteRecord", nil, in, nil)
}

// ---- write verbs ----

// Favorite favorites a station. Idempotent: the record key is derived from the
// station id (FavoriteRkey), so favoriting twice overwrites one record. It also
// reconciles favorites saved under old random keys — after writing the canonical
// record it deletes any other favorite for the same stationId (matched on the
// record body). Returns the canonical record URI.
func (a *Agent) Favorite(ctx context.Context, station StationInfo) (string, error) {
	canonical := FavoriteRkey(station.StationID)
	rec := favoriteRecord{Type: NSIDFavorite, Station: station, CreatedAt: nowRFC3339()}
	if err := a.putRecord(ctx, NSIDFavorite, canonical, rec); err != nil {
		return "", err
	}
	if rkeys, err := a.favoriteRkeysFor(ctx, station.StationID); err == nil {
		for _, rkey := range rkeys {
			if rkey != canonical {
				_ = a.deleteRecord(ctx, NSIDFavorite, rkey) // best-effort
			}
		}
	}
	return fmt.Sprintf("at://%s/%s/%s", a.did, NSIDFavorite, canonical), nil
}

// Unfavorite deletes every favorite record for the station's stationId.
func (a *Agent) Unfavorite(ctx context.Context, station StationInfo) error {
	rkeys, err := a.favoriteRkeysFor(ctx, station.StationID)
	if err != nil {
		return err
	}
	for _, rkey := range rkeys {
		if err := a.deleteRecord(ctx, NSIDFavorite, rkey); err != nil {
			return err
		}
	}
	return nil
}

// Comment posts a comment on a station. Returns the record URI.
func (a *Agent) Comment(ctx context.Context, station StationInfo, text string) (string, error) {
	rec := commentRecord{Type: NSIDComment, Station: station, Text: text, CreatedAt: nowRFC3339()}
	return a.createRecord(ctx, NSIDComment, rec)
}

// CreateStation creates a custom station. Returns the record URI.
func (a *Agent) CreateStation(ctx context.Context, draft StationDraft) (string, error) {
	rec := stationRecord{
		Type:      NSIDStation,
		Name:      draft.Name,
		StreamURL: draft.StreamURL,
		Genre:     draft.Genre,
		Homepage:  draft.Homepage,
		Logo:      draft.Logo,
		CreatedAt: nowRFC3339(),
	}
	return a.createRecord(ctx, NSIDStation, rec)
}

// SetPlayStatus updates the actor's play-status singleton (rkey `self`).
func (a *Agent) SetPlayStatus(ctx context.Context, station StationInfo) error {
	rec := actorStatusRecord{Type: NSIDActorStatus, Station: station, PlayedAt: nowRFC3339()}
	return a.putRecord(ctx, NSIDActorStatus, selfRkey, rec)
}

// DeletePlayStatus removes the actor's play-status singleton. Idempotent.
func (a *Agent) DeletePlayStatus(ctx context.Context) error {
	if err := a.deleteRecord(ctx, NSIDActorStatus, selfRkey); err != nil && !isNotFound(err) {
		return err
	}
	return nil
}

// GetAudioSettings fetches the synced audio-settings singleton, or nil if none
// exists yet (first run on this account).
func (a *Agent) GetAudioSettings(ctx context.Context) (*AudioSettings, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	var out struct {
		Value AudioSettings `json:"value"`
	}
	params := map[string]any{"repo": a.did, "collection": NSIDAudioSettings, "rkey": selfRkey}
	if err := a.xrpc.Do(ctx, xrpc.Query, "", "com.atproto.repo.getRecord", params, nil, &out); err != nil {
		if isNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	return &out.Value, nil
}

// PutAudioSettings upserts the audio-settings singleton (rkey `self`).
func (a *Agent) PutAudioSettings(ctx context.Context, settings AudioSettings) error {
	settings.Type = NSIDAudioSettings
	settings.UpdatedAt = nowRFC3339()
	return a.putRecord(ctx, NSIDAudioSettings, selfRkey, settings)
}

// MintServiceAuth mints an atproto service-auth JWT bound to aud (the AppView's
// DID service reference) and lxm (the lexicon method).
func (a *Agent) MintServiceAuth(ctx context.Context, aud, lxm string) (string, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	var out struct {
		Token string `json:"token"`
	}
	params := map[string]any{"aud": aud, "lxm": lxm, "exp": time.Now().Unix() + 60}
	if err := a.xrpc.Do(ctx, xrpc.Query, "", "com.atproto.server.getServiceAuth", params, nil, &out); err != nil {
		return "", err
	}
	return out.Token, nil
}

// favoriteRkeysFor returns the rkeys of every favorite whose body stationId
// matches. The rkey isn't the join key (schemes differ), so we read the
// stationId out of each record. Paginates the repo.
func (a *Agent) favoriteRkeysFor(ctx context.Context, stationID string) ([]string, error) {
	var rkeys []string
	cursor := ""
	for {
		a.mu.Lock()
		var out struct {
			Cursor  string `json:"cursor"`
			Records []struct {
				URI   string `json:"uri"`
				Value struct {
					Station struct {
						StationID string `json:"stationId"`
					} `json:"station"`
				} `json:"value"`
			} `json:"records"`
		}
		params := map[string]any{"repo": a.did, "collection": NSIDFavorite, "limit": 100}
		if cursor != "" {
			params["cursor"] = cursor
		}
		err := a.xrpc.Do(ctx, xrpc.Query, "", "com.atproto.repo.listRecords", params, nil, &out)
		a.mu.Unlock()
		if err != nil {
			return nil, err
		}
		for _, r := range out.Records {
			if r.Value.Station.StationID == stationID {
				rkeys = append(rkeys, rkeyFromURI(r.URI))
			}
		}
		if out.Cursor == "" || len(out.Records) == 0 {
			break
		}
		cursor = out.Cursor
	}
	return rkeys, nil
}

func isNotFound(err error) bool {
	return err != nil && strings.Contains(err.Error(), "RecordNotFound")
}

File: sdk/typescript/src/agent.ts
import "@atcute/atproto"; // side-effect: augments the ambient com.atproto.* XRPC maps
import {
  Client,
  ClientResponseError,
  CredentialManager,
  ok,
  type AtpSessionData,
} from "@atcute/client";
import type { Did, Nsid } from "@atcute/lexicons";
import { now as tidNow } from "@atcute/tid";
import {
  ACTOR_STATUS_RKEY,
  AUDIO_SETTINGS_RKEY,
  audioSettingsRecordToData,
  buildActorStatusRecord,
  buildAudioSettingsRecord,
  buildCommentRecord,
  buildFavoriteRecord,
  buildStationRecord,
  NSID,
  rkeyFromUri,
  stationRecordToStation,
  type AudioSettingsData,
  type AudioSettingsRecord,
  type FavoriteRecord,
  type GifEmbed,
  type Mention,
  type Station,
  type StationDraft,
} from "@atradio/lexicons";
import { AppView } from "./appview.ts";
import { favoriteRkey } from "./favorite-rkey.ts";

/** Cast a record object to the shape atcute's putRecord expects. */
function asRecord(value: object): Record<string, unknown> {
  return value as unknown as Record<string, unknown>;
}

export interface LoginOptions {
  /** The account's PDS (or an entryway like `https://bsky.social`). */
  service?: string;
  identifier: string;
  password: string;
  /** Email 2FA code, if enabled. */
  code?: string;
  /** AppView base URL (defaults to `https://api.atradio.fm`). */
  appview?: string;
  /** Called whenever the session is created/refreshed — persist it here. */
  onSessionUpdate?: (session: AtpSessionData) => void;
}

export interface AgentOptions {
  appview?: string;
}

/**
 * The atradio.fm agent, mirroring the Rust SDK's `AtradioAgent`: a thin wrapper
 * over an atcute {@link Client} that exposes high-level record verbs plus a
 * read-only {@link AppView}.
 *
 * Construct it with app-password auth via {@link AtradioAgent.login}, or wrap an
 * already-authenticated client (e.g. the web's OAuth session) with
 * {@link AtradioAgent.fromClient}.
 */
export class AtradioAgent {
  readonly client: Client;
  readonly did: Did;
  readonly appview: AppView;
  readonly #manager?: CredentialManager;

  constructor(
    client: Client,
    did: Did,
    opts: AgentOptions & { manager?: CredentialManager } = {},
  ) {
    this.client = client;
    this.did = did;
    this.appview = new AppView(opts.appview);
    this.#manager = opts.manager;
  }

  /** Log in with an app password and return a ready agent. */
  static async login(opts: LoginOptions): Promise<AtradioAgent> {
    const manager = new CredentialManager({
      service: opts.service ?? "https://bsky.social",
      onSessionUpdate: opts.onSessionUpdate,
    });
    const session = await manager.login({
      identifier: opts.identifier,
      password: opts.password,
      code: opts.code,
    });
    const client = new Client({ handler: manager });
    return new AtradioAgent(client, session.did, {
      appview: opts.appview,
      manager,
    });
  }

  /** Wrap an already-authenticated atcute client (e.g. an OAuth session). */
  static fromClient(
    client: Client,
    did: Did,
    opts: AgentOptions = {},
  ): AtradioAgent {
    return new AtradioAgent(client, did, opts);
  }

  /** The current app-password session, if this agent owns one. */
  get session(): AtpSessionData | undefined {
    return this.#manager?.session;
  }

  /**
   * Keep the session alive by touching an authenticated endpoint, which makes
   * the handler refresh the token if it's near/at expiry. Call on a timer (and
   * once at startup) so writes keep working after a long idle. Rejects only if
   * the session is unrecoverable (re-login needed).
   */
  async refreshSession(): Promise<void> {
    await ok(this.client.get("com.atproto.server.getSession", {}));
  }

  // ---- writes ----------------------------------------------------------

  /**
   * Favorite a station. Idempotent: the record key is derived from the station
   * id ({@link favoriteRkey}), so favoriting twice overwrites one record. Also
   * reconciles favorites saved under the old random keys — after writing the
   * canonical record it deletes any other favorite for the same `stationId`
   * (matched on the record body, since keys differ). Returns the record URI.
   */
  async favorite(station: Station): Promise<string> {
    const rkey = await favoriteRkey(station.id);
    await ok(
      this.client.post("com.atproto.repo.putRecord", {
        input: {
          repo: this.did,
          collection: NSID.favorite as Nsid,
          rkey,
          record: asRecord(buildFavoriteRecord(station)),
        },
      }),
    );
    await this.#pruneFavorites(station.id, rkey);
    return `at://${this.did}/${NSID.favorite}/${rkey}`;
  }

  /** Unfavorite a station: delete every favorite record for its `stationId`. */
  async unfavorite(station: Station): Promise<void> {
    await this.#pruneFavorites(station.id);
  }

  /** Post a comment on a station. Returns the record URI. */
  async comment(
    station: Station,
    text: string,
    opts: { facets?: Mention[]; gif?: GifEmbed } = {},
  ): Promise<string> {
    const rkey = tidNow();
    await ok(
      this.client.post("com.atproto.repo.putRecord", {
        input: {
          repo: this.did,
          collection: NSID.comment as Nsid,
          rkey,
          record: asRecord(buildCommentRecord(station, text, opts)),
        },
      }),
    );
    return `at://${this.did}/${NSID.comment}/${rkey}`;
  }

  /** Create a custom station. Returns its rkey + the resulting `Station`. */
  async createStation(
    draft: StationDraft,
  ): Promise<{ rkey: string; station: Station }> {
    const rkey = tidNow();
    const record = buildStationRecord(draft);
    await ok(
      this.client.post("com.atproto.repo.putRecord", {
        input: {
          repo: this.did,
          collection: NSID.station as Nsid,
          rkey,
          record: asRecord(record),
        },
      }),
    );
    return { rkey, station: stationRecordToStation(record, rkey) };
  }

  /** Update the actor's play-status singleton (rkey `self`). */
  async setPlayStatus(station: Station): Promise<void> {
    await ok(
      this.client.post("com.atproto.repo.putRecord", {
        input: {
          repo: this.did,
          collection: NSID.actorStatus as Nsid,
          rkey: ACTOR_STATUS_RKEY,
          record: asRecord(buildActorStatusRecord(station)),
        },
      }),
    );
  }

  /** Delete the actor's play-status singleton. Idempotent. */
  async deletePlayStatus(): Promise<void> {
    try {
      await ok(
        this.client.post("com.atproto.repo.deleteRecord", {
          input: {
            repo: this.did,
            collection: NSID.actorStatus as Nsid,
            rkey: ACTOR_STATUS_RKEY,
          },
        }),
      );
    } catch (err) {
      if (!isRecordNotFound(err)) throw err;
    }
  }

  /** Fetch the synced audio settings, or `null` if none saved yet. */
  async getAudioSettings(): Promise<AudioSettingsData | null> {
    try {
      const out = await ok(
        this.client.get("com.atproto.repo.getRecord", {
          params: {
            repo: this.did,
            collection: NSID.audioSettings as Nsid,
            rkey: AUDIO_SETTINGS_RKEY,
          },
        }),
      );
      return audioSettingsRecordToData(
        out.value as unknown as AudioSettingsRecord,
      );
    } catch (err) {
      if (isRecordNotFound(err)) return null;
      throw err;
    }
  }

  /** Upsert the audio-settings singleton (rkey `self`). */
  async putAudioSettings(data: AudioSettingsData): Promise<void> {
    await ok(
      this.client.post("com.atproto.repo.putRecord", {
        input: {
          repo: this.did,
          collection: NSID.audioSettings as Nsid,
          rkey: AUDIO_SETTINGS_RKEY,
          record: asRecord(buildAudioSettingsRecord(data)),
        },
      }),
    );
  }

  /**
   * Mint an atproto service-auth JWT bound to `aud` (the AppView's DID service
   * reference) and `lxm` (the lexicon method) — proves to the atradio Connect
   * hub that a connection belongs to this account.
   */
  async mintServiceAuth(aud: string, lxm: string): Promise<string> {
    const out = await ok(
      this.client.get("com.atproto.server.getServiceAuth", {
        params: {
          aud: aud as Did,
          lxm: lxm as Nsid,
          exp: Math.floor(Date.now() / 1000) + 60,
        },
      }),
    );
    return out.token;
  }

  // ---- internals -------------------------------------------------------

  /** Every favorite record whose body `station.stationId` matches. */
  async #favoriteRkeysFor(stationId: string): Promise<string[]> {
    const rkeys: string[] = [];
    let cursor: string | undefined;
    do {
      const page = await ok(
        this.client.get("com.atproto.repo.listRecords", {
          params: {
            repo: this.did,
            collection: NSID.favorite as Nsid,
            limit: 100,
            cursor,
          },
        }),
      );
      for (const r of page.records) {
        const rec = r.value as unknown as FavoriteRecord;
        if (rec.station?.stationId === stationId) {
          rkeys.push(rkeyFromUri(r.uri));
        }
      }
      cursor = page.cursor;
    } while (cursor);
    return rkeys;
  }

  /** Delete every favorite for `stationId` except `keep` (best-effort). */
  async #pruneFavorites(stationId: string, keep?: string): Promise<void> {
    const rkeys = await this.#favoriteRkeysFor(stationId);
    for (const rkey of rkeys) {
      if (rkey === keep) continue;
      await ok(
        this.client.post("com.atproto.repo.deleteRecord", {
          input: {
            repo: this.did,
            collection: NSID.favorite as Nsid,
            rkey,
          },
        }),
      );
    }
  }
}

function isRecordNotFound(err: unknown): boolean {
  return err instanceof ClientResponseError && err.error === "RecordNotFound";
}
